accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,140 @@
+	overrideAddrs map[string]common.Address
+	deployedAddrs map[string]common.Address
+	deployerTxs   map[string]*types.Transaction
+	input         map[string][]byte // map of the root contract pattern to the constructor input (if there is any)
+	deploy        DeployFn
+}
+
+// linkAndDeploy recursively deploys a contract and its dependencies:  starting by linking/deploying its dependencies.
+// The deployment result (deploy addresses/txs or an error) is stored in the depTreeDeployer object.
+func (d *depTreeDeployer) linkAndDeploy(metadata *MetaData) error {
+	// don't deploy contracts specified as overrides.  don't deploy their dependencies.
+	if _, ok := d.overrideAddrs[metadata.Pattern]; ok {
+		return nil
+	}
+	// if this contract/library depends on other libraries deploy them (and their dependencies) first
+	for _, dep := range metadata.Deps {
+		if err := d.linkAndDeploy(dep); err != nil {
+			return err
+		}
+	}
+	// if we just deployed any prerequisite contracts, link their deployed addresses into the bytecode to produce
+	// a deployer bytecode for this contract.
+	deployerCode := metadata.Bin
+	for _, dep := range metadata.Deps {
+		linkAddr, ok := d.deployedAddrs[dep.Pattern]
+		if !ok {
+			linkAddr = d.overrideAddrs[dep.Pattern]
+		}
+		deployerCode = strings.ReplaceAll(deployerCode, "__$"+dep.Pattern+"$__", strings.ToLower(linkAddr.String()[2:]))
+	}
+
+	// Finally, deploy the contract.
+	deployer, err := hex.DecodeString(deployerCode[2:])
+	if err != nil {
+		panic(fmt.Sprintf("error decoding contract deployer hex %s:\n%v", deployerCode[2:], err))
+	}
+	addr, tx, err := d.deploy(d.input[metadata.Pattern], deployer)
+	if err != nil {
+		return err
+	}
+
+	d.deployedAddrs[metadata.Pattern] = addr
+	d.deployerTxs[metadata.Pattern] = tx
+	return nil
+}
+
+// result returns a result for this deployment, or an error if it failed.
+func (d *depTreeDeployer) result() *DeploymentResult {
+	return &DeploymentResult{
+		Txs:   d.deployerTxs,
+		Addrs: d.deployedAddrs,
+	}
+}
+
+func newDepTreeDeployer(overrides map[string]common.Address, deploy DeployFn) *depTreeDeployer {
+	return &depTreeDeployer{
+		deploy:        deploy,
+		overrideAddrs: overrides,
+		deployedAddrs: make(map[string]common.Address),
+		deployerTxs:   make(map[string]*types.Transaction)}
+}
+
+// LinkAndDeploy deploys a specified set of contracts and their dependent
+// libraries.  If an error occurs, only contracts which were successfully
+// deployed are returned in the result.
+func LinkAndDeploy(deployParams *DeploymentParams, deploy DeployFn) (res *DeploymentResult, err error) {
+	accumRes := &DeploymentResult{
+		Txs:   make(map[string]*types.Transaction),
+		Addrs: make(map[string]common.Address),
+	}
+	deployer := newDepTreeDeployer(deployParams.overrides, deploy)
+	for _, contract := range deployParams.contracts {
+		if deployParams.inputs != nil {
+			deployer.input = map[string][]byte{contract.Pattern: deployParams.inputs[contract.Pattern]}
+		}
```

Comment:
This clause is not covered by testing (commenting it out doesn't cause failures). And it's not surprising, because the clause doesn't make sense. 
On each iteration of a `contract`, it overwrites the `input`-map with the current `contract` input. 
Why not just copy/clone the `deployParams.inputs` map  to the `deployer`, already when creating the `deployer`?

Reply:
Afaict you can turn it into 
```golang
func LinkAndDeploy(deployParams *DeploymentParams, deploy DeployFn) (res *DeploymentResult, err error) {
	deployer := newDepTreeDeployer(deployParams.overrides, deploy, deployParams.inputs)
	for _, contract := range deployParams.contracts {
		if _ err := deployer.linkAndDeploy(contract); err != nil {
			return deployer.result(), err
		}
	}
	return deployer.result(), nil
```